### PR TITLE
Display status and apply resources synchronously

### PR DESCRIPTION
**Pull request by aolfat**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: client/src/featureform/register.py [hunk 1]
```diff
@@ -2878,12 +2878,82 @@ def apply(self):
...
+        if self.local:
+            state().create_all_local()
+        else:
+            resource_state = state()
+            resource_state.create_all(self._stub)
+            resources = resource_state.sorted_list()
+            while True:
+                # get the status for each resource so we can display it
+                statuses = self.get_statuses(resources)
+                self.display_status(statuses)
```
Comment (simba-git) [inline client/src/featureform/register.py:2895]: The time.sleep should probably get pulled up into here if you need it, this also needs a break

File: client/src/featureform/cli.py [hunk 2]
```diff
@@ -173,6 +173,9 @@ def dash():
 @click.option("--dry-run",
               is_flag=True,
               help="Checks the definitions without applying them")
+@click.option("--poll-status",
```
Comment (simba-git) [inline client/src/featureform/cli.py:176]: This should probably be --wait or --async, --poll-status seems pretty non-standard and is focused on the implementation

File: client/src/featureform/resources.py [hunk 3]
```diff
@@ -1322,6 +1428,29 @@ def operation_type() -> OperationType:
...
+            name=ts.name,
+            variant=ts.variant,
+            owner=ts.owner,
+            description=ts.description,
+            status=ts.status.Status._enum_type.values[ts.status.status].name,
+            label=(ts.label.name, ts.label.variant),
+            features=[(f.name, f.variant) for f in ts.features],
+            feature_lags=[],
+            provider=ts.provider,
+            # TODO: apply values from proto
```
Comment (simba-git) [inline client/src/featureform/resources.py:1448]: ?

Comment (aolfat) [reply client/src/featureform/resources.py:1448]: That was already there haha

File: client/src/featureform/cli.py (hunk 2 above)
Comment (aolfat) [reply client/src/featureform/cli.py:176]: Will change to wait

Comment (simba-git) [reply client/src/featureform/cli.py:176]: Cool, you haven't done yet so unresolving so we don't forget

File: client/src/featureform/format.py [hunk 4]
```diff
@@ -20,6 +31,98 @@ def format_rows(format_obj, format_obj_2=None, format_obj_3=None, format_obj_4=N
...
+            table.add_column("Status", width=10)
+            table.add_column("Error", style="red")
+
+            for status in statuses:
+                error = f" {status.error}" if status.error else ""
+                resource_type = status.resource_type.__name__
+                name = status.name
+                status_text = status.status
+
+                if status.resource_type is Provider:
```
Comment (simba-git) [inline client/src/featureform/format.py:102]: Could change this whole code block to:
status_text = status.status if status.resource_type is not Provider else "CREATED"
text_color = {
  "CREATED": "green",
  ...
}
status_type = text_color[status_text]

File: client/src/featureform/format.py [hunk 5]
```diff
@@ -20,6 +31,98 @@ def format_rows(format_obj, format_obj_2=None, format_obj_3=None, format_obj_4=N
...
+                    if status.status == "NO_STATUS":
```
Comment (simba-git) [inline client/src/featureform/format.py:103]: This is always true right? If statement not needed here

File: client/src/featureform/resources.py [hunk 6]
```diff
@@ -667,19 +666,22 @@ def __post_init__(self):
...
 @dataclass
 class Provider:
     name: str
-    function: str
-    config: Config
     description: str
     team: str
-    tags: list
-    properties: dict
+    config: Optional[Config] = None  # TODO remove optional and add deserializer to configs
```
Comment (simba-git) [inline client/src/featureform/resources.py:676]: Can we resolve this TODO easily, I don't like adding TODOs in code because they tend to never get done

Comment (simba-git) [reply client/src/featureform/resources.py:676]: We can do it early next week, not worth blocking the release on obviously but lets knock it out soon

File: client/src/featureform/resources.py [hunk 7]
```diff
@@ -1322,6 +1428,29 @@ def operation_type() -> OperationType:
...
+            variant=ts.variant,
+            owner=ts.owner,
+            description=ts.description,
+            status=ts.status.Status._enum_type.values[ts.status.status].name,
+            label=(ts.label.name, ts.label.variant),
+            features=[(f.name, f.variant) for f in ts.features],
+            feature_lags=[],
+            provider=ts.provider,
+            # TODO: apply values from proto
+            tags=[],
```
Comment (simba-git) [inline client/src/featureform/resources.py:1449]: Shouldn't we set tags and properties? Ok if not easy to do but lets make note of it so we actually do it

**Review by simba-git — APPROVED** [review]

File: client/src/featureform/cli.py (hunk 2 above)
Comment (sdreyer) [reply client/src/featureform/cli.py:176]: Also rn is seems like this flag doesnt do anything. Should have it set wait=True if the flag is passed

File: client/src/featureform/format.py (hunk 4 above)
Comment (aolfat) [reply client/src/featureform/format.py:102]: good call. Yeah I didn't refactor what chatgpt spat out

File: client/src/featureform/resources.py (hunk 7 above)
Comment (aolfat) [reply client/src/featureform/resources.py:1449]: we should. I can do it as a separate PR

File: client/src/featureform/format.py [hunk 8]
```diff
@@ -20,6 +30,103 @@ def format_rows(format_obj, format_obj_2=None, format_obj_3=None, format_obj_4=N
...
 
+
 def format_pg(s=""):
     print(divider)
     print(s)
+
+
+def display_statuses(stub, resources):
+
+    from featureform import Feature, TrainingSet, Label, Source, Provider
```
Comment (aolfat) [inline client/src/featureform/format.py:41]: to avoid cyclic imports

**Review by sdreyer — APPROVED** [review]

**Pull request merged**